Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h

```diff
@@ -0,0 +1,363 @@
+    };
+
+    void setup_graph(const Filtered_dual_graph& dual)
+    {
+        std::map<face_descriptor, graph_vertex_descriptor> face_graph_map;
+
+        m_cluster_map = boost::get(cluster_props_t(), m_graph);
+        m_decimation_map = boost::get(decimation_props_t(), m_graph);
+
+        BOOST_FOREACH(face_descriptor face, vertices(dual))
+        {
+            Cluster_properties props;
+            
+            props.concavity = 0;
+            props.faces.push_back(face);
+
+            BOOST_FOREACH(vertex_descriptor vert, CGAL::vertices_around_face(CGAL::halfedge(face, m_mesh), m_mesh))
+            {
+                props.conv_hull_pts.push_back(m_mesh.point(vert));
+            }
+
+            face_graph_map[face] = add_vertex(props, m_graph);
+        }
+
+        BOOST_FOREACH(edge_descriptor edge, edges(dual))
+        {
+            Decimation_properties props;
+
+            add_edge(face_graph_map[source(edge, dual)], face_graph_map[target(edge, dual)], props, m_graph);
+        }
+    }
+
+    void update_edge(graph_edge_descriptor edge, double concavity_threshold, double alpha_factor)
+    {
+//        std::cout << std::endl << "Update edge: " << edge << std::endl;
+
+        Decimation_properties& decimation_props = m_decimation_map[edge];
+
+        graph_vertex_descriptor vert_1 = source(edge, m_graph), vert_2 = target(edge, m_graph);
+
+        Cluster_properties& cluster_1_props = m_cluster_map[vert_1];
+        Cluster_properties& cluster_2_props = m_cluster_map[vert_2];
+
+        std::vector<Point_3> common_hull_pts;
+        
+        CGAL_assertion(cluster_1_props.conv_hull_pts.size() >= 3);
+        CGAL_assertion(cluster_2_props.conv_hull_pts.size() >= 3);
+        
+        BOOST_FOREACH(Point_3 p, cluster_1_props.conv_hull_pts)
+        {
+            common_hull_pts.push_back(p);
+        }
+        BOOST_FOREACH(Point_3 p, cluster_2_props.conv_hull_pts)
+        {
+            common_hull_pts.push_back(p);
+        }
+
+        Surface_mesh conv_hull;
+        CGAL::convex_hull_3(common_hull_pts.begin(), common_hull_pts.end(), conv_hull);
+
+        decimation_props.new_cluster_props.conv_hull_pts.clear();
+        BOOST_FOREACH(vertex_descriptor vert, CGAL::vertices(conv_hull))
+        {
+            decimation_props.new_cluster_props.conv_hull_pts.push_back(conv_hull.point(vert));
+        }
+        CGAL_warning(decimation_props.new_cluster_props.conv_hull_pts.size() > 3);
+
+        decimation_props.new_cluster_props.faces.clear();
+        BOOST_FOREACH(face_descriptor face, cluster_1_props.faces)
+        {
+            decimation_props.new_cluster_props.faces.push_back(face);
+        }
+        BOOST_FOREACH(face_descriptor face, cluster_2_props.faces)
+        {
+            decimation_props.new_cluster_props.faces.push_back(face);
+        }
+
+        Surface_mesh new_cluster;
+        Face_filtered_mesh selected_faces(m_mesh, decimation_props.new_cluster_props.faces);
```

Comment:
why do you need a Face_filtered_graph here? I'm surprized it compiles with the default Polyhedron type?

Reply:
Each cluster contains a list of faces that corresponds to the cluster. I construct new mesh using that list and then use it for concavity computation. It is not efficient and needs to be improved.

Reply:
First, you don't need to copy the mesh, the `Face_filtered_graph` could be used directly in `Concavity`. Then you might even get rid of it by simply collecting the vertices per cluster if this is faster.

Reply:
Actually you should precompute the convex hull to avoid computing it over and over while calling `update_edge` in your the for loops.

Reply:
Good point about not copying. I need that mesh in concavity.h for normals computation. To compute the normals you need to know adjacent faces of a vertex.

update_edge is called only when any of two smaller clusters is changed. It computes all the necessary parameters like decimation cost, concavity, and later they are used for choosing next optimal edge without again calling update_edge, which implies computation of convex hull.
I think about maintaining dynamic convex hulls or implementing custom computation of convex hull using the fact that it consists of two smaller convex hulls rather than a list of random points.

Do you know any tools for profiling the code (gcc, linux)?

Reply:
What I mean, if you might don't need to create the filtered_face_graph and only use the input mesh with the selection of vertices. About convex hull, you can start by considering the points on the convex hull of step n-1 + those entering the cluster (this will already filter out some useless points). `gprof` for the profiling. There should also be some nice gui on top of it.

Reply:
1. Okay, I understand you.
2. I already do so :)
3. Thanks.